File: src/main/java/eightyDays/java8/bank/partner/Partner.java
package eightyDays.java8.bank.partner;

import eightyDays.java8.bank.Bank;

import java.math.BigDecimal;
import java.util.function.Predicate;

public abstract class Partner {
    private String name;

    public Partner(String pName) {
        name = pName;
    }

    public String getName() {
        return name;
    }

    public BigDecimal asset(Bank bank) {
        // TODO
        return BigDecimal.ONE;
    }

    public static Predicate<Partner> byName(String name) {
        return partner -> partner.getName().equals(name);
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;

        Partner partner = (Partner) o;

        return name.equals(partner.name);

    }

    @Override
    public int hashCode() {
        return name.hashCode();
    }
}

File: src/main/scala/eightyDays/scala211/bank/Partner.scala
package eightyDays.scala211.bank

import eightyDays.scala211.bank.account.{Account, Amount}

  abstract class Partner(val name: String) {
    def asset(bank: Bank): Amount = bank.filterAccounts(Account.byOwner(this)).foldLeft(BigDecimal.valueOf(0)) { case (balance, account) => balance + account.balance }
  }

  object Partner {
    def byName(name: String): Partner => Boolean = partner => partner.name == name
  }

  case class Person(firstName: String, override val name: String) extends Partner(name)
  case class LegalEntity(override val name: String) extends Partner(name)

  object Person {
    def byFirstname(firstName: String): Partner => Boolean = {
      case Person(`firstName`, _) => true
      case _ => false
    }
  }

  case class Identification(number: String = java.util.UUID.randomUUID.toString)